L69-cpp-core-event-manager.md

```diff
@@ -0,0 +1,282 @@
+
+## Proposal
+The following event manager interfaces will be available in the gRPC core.
+
+```cpp
+// An external handle to an EventManager task, which can be used to cancel the 
+// task. The EventManager task is added via RunFunction.
+struct TaskKey;
+
+// An external handle to an EventManager alarm, which can be used to cancel the
+// alarm. The EventManager alarm is added via AddAlarmIn.
+struct AlarmKey;
+
+// Scheduling priority for closures and alarms added into the EventManager.
+enum Priority {
+  PRIORITY_MIN = 0,
+  PRIORITY_LOW = 0,
+  PRIORITY_MED = 1,
+  PRIORITY_HIGH = 2,
+  PRIORITY_IMMEDIATE = 3,
+  PRIORITY_SYSTEM = 4,
+  PRIORITY_MAX = 4,
+  NUM_PRIORITIES = 5
+};
+
+// A top-level base class of EventManagerInterface.
+class BaseEventManagerInterface {
+ public:
+  virtual ~BaseEventManagerInterface();
+
+  // Adds a new one-shot alarm, which will be scheduled at the deadline to run the 
+  // closure with true. The deadline here is an absolute time. The AlarmKey pointer 
+  // is an output parameter which can be used to cancel the alarm before it fires. 
+  // Note that the closure should not be blocking or long-running, as it would 
+  // block the worker thread in the EventManager from polling the file 
+  // descriptors. In case that the alarm is cancelled successfully in 
+  // ConsistentCancel, the closure will be run with false in the calling thread of 
+  // ConsistentCancel. Callers of ConsistentCancel need to be cautious to avoid 
+  // deadlock.
+  virtual void ScheduleAlarmIn(gpr_timespec deadline, std::function<void(bool)> c, Priority priority, AlarmKey* key = nullptr) = 0;
+
+  // Adds a closure to the EventManager, which will be run at the earliest possible 
+  // time. The TaskKey pointer is an output parameter which can be used to cancel 
+  // the closure at a later time (currently not supported in 
+  // BaseEventManagerInterface). Note that the closure should not be blocking or 
+  // long-running, as it would block the worker thread in the EventManager from 
+  // polling the file descriptors.
+  virtual void RunFunction(std::function<void()> handler, Priority priority, TaskKey* key = nullptr) = 0;
+
+  struct Task {
+    std::function<void()> handler;
+    Priority priority;
+    TaskKey* key;
+    int internal_added_queue;
+  };
+  // Adds multiple closures to the event management system all at once.
+  virtual void RunFunctions(Task* tasks, int length) = 0;
+
+  // Cancels an alarm which has been added to the EventManager via 
+  // ScheduleAlarmIn(). This function will run the closure with false in the 
+  // calling thread if it successfully cancels the alarm.
+  virtual void ConsistentCancel(const AlarmKey& key) = 0;
```

Comment:
Can you explain the use of the word "Consistent" in this member function name?

Reply:
It is to differentiate from Google's implementation of `Cancel` in EM where canceling an alarm will not trigger the closure to run (in other words, `ConsistentCancel` will consistently invoke the alarm closure despite that the alarm is cancelled). I'll update the comments.

Reply:
Comment updates look good. Thanks.

As a non-Google user who's never seen the Google EM implementation, the word "Consistent" in the name still comes off as a bit odd. For me, just the comments and "Cancel" would be enough. I encourage you to balance how much internal Google bits "leak", if you will, into this design.

(If this interface had both behaviors, that would be a different story, but it only has one behavior.)

Reply:
Agree that the name is odd and agree that we will try to minimize 'internal leak' in this design. However in this case a leak is inevitable and we can't simply name the method as `Cancel`. I can rename it to something more sensible like `CancelAndTrigger`. Suggestion welcomed.

Reply:
Maybe `CancelAndRun` for parallelism with `RunFunction`? ("Trigger" doesn't show up in the API names or comments yet. "Run" does.)

Is conflict trying to be avoided a member function name clash, a conceptual/semantic clash, or something else? I'm still a bit fuzzy here.

Reply:
`CancelAndRun` is good.

The conflict is a semantics clash as Google's EM is going to inherit from this interface, so if we name it `Cancel` the semantics is different between the interface and Google's implementation.

Reply:
One thing I would like to call out here is that this function won't run the closure if the alarm has completed already (i.e. we fail to cancel the alarm).

On the other hand, the new name `CancelAndRun` looks good to me.